### PR TITLE
Add deframe op and stft/istft api.

### PR DESCRIPTION
**Pull request by KPatr1ck**
- PR types
New features

- PR changes
OPs: deframe
Apis: deframe, stft, istft

- Describe
  - Add Seq2Col functor. 
  - Add deframe op and CPU/GPU kernels.
  - Add deframe, stft and istft api.

### CODE REVIEW

**Comment by KPatr1ck** [comment]
# Testing script

```python
import ipdb
import librosa
import numpy as np
import paddle
from paddle.tensor.signal import stft, istft
from paddle.tensor.fft import fft, rfft, irfft
from scipy.signal.windows import get_window
import torch

np_type = None


def switch_float_type(t: int=64):
    global np_type
    if t == 16:
        np_type = np.float16
    elif t == 32:
        np_type = np.float32
    elif t == 64:
        np_type = np.float64
    else:
        raise ValueError(f'Float type shoulb be one of [16, 32, 64], but got {t}')


# ==== 1. stft ====
def test_stft_api(shape, n_fft, hop_length, window, device='cpu'):
    print('='*30, '1. [API]stft', '='*30)
    paddle.set_device(device)

    if window is None:
        win_p = None
        win_t = None
        win_l = 'boxcar'  # rectangular window
    elif isinstance(window, np.ndarray):
        win_p = paddle.to_tensor(window)
        win_t = torch.from_numpy(window)
        win_l = window
    else:
        raise ValueError(f'window should be None or an ndarray, but got {type(window)}')

    x = np.random.random(size=shape).astype(np_type)

    xp = paddle.to_tensor(x)
    xt = torch.from_numpy(x)

    yp = stft(xp, n_fft=n_fft, hop_length=hop_length, window=win_p)
    yt = torch.stft(xt, n_fft=n_fft, hop_length=hop_length, return_complex=True, window=win_t)
    yl = librosa.stft(x.squeeze(0), n_fft=n_fft, hop_length=hop_length, window=win_l)  # only 1D input support.

    print('*'*10, 'in dtype', '*'*10)
    print(f'paddle: {xp.dtype}, torch: {xt.dtype}, librosa: {x.dtype}')
    print('*'*10, 'out dtype', '*'*10)
    print(f'paddle: {yp.dtype}, torch: {yt.dtype}, librosa: {yl.dtype}')
    print()

    print('*'*10, 'in shape', '*'*10)
    print(f'paddle: {xp.shape}, torch: {xt.shape}, librosa: {x.shape}')
    print('*'*10, 'out shape', '*'*10)
    print(f'paddle: {yp.shape}, torch: {yt.shape}, librosa: {yl.shape}')
    print()

    print('*'*10, 'torch', '*'*10)
    for rtol in [1.e-7, 1.e-6, 1.e-5, 1.e-4, 1.e-3]:
        print(f'[paddle vs torch  ][rtol={rtol:.1E}]: {np.allclose(yp.numpy(), yt.numpy(), rtol=rtol)}')

    print('*'*10, 'librosa', '*'*10)
    for rtol in [1.e-7, 1.e-6, 1.e-5, 1.e-4, 1.e-3]:
        print(f'[paddle vs librosa][rtol={rtol:.1E}]: {np.allclose(yp.numpy(), np.expand_dims(yl, 0), rtol=rtol)}')

    print('*'*10, 'torch vs librosa', '*'*10)
    for rtol in [1.e-7, 1.e-6, 1.e-5, 1.e-4, 1.e-3]:
        print(f'[torch  vs librosa][rtoj={rtol:.1E}]: {np.allclose(yt.numpy(), np.expand_dims(yl, 0), rtol=rtol)}')

    return yp, yt, yl

# ==== 2. istft ====
def test_istft_api(y, n_fft, hop_length, window, device='cpu', **kwargs):
    print('='*30, '2. [API]istft', '='*30)
    paddle.set_device(device)

    if window is None:
        win_p = None
        win_t = None
        win_l = 'boxcar'  # rectangular window
    elif isinstance(window, np.ndarray):
        win_p = paddle.to_tensor(window)
        win_t = torch.from_numpy(window)
        win_l = window
    else:
        raise ValueError(f'window should be None or an ndarray, but got {type(window)}')

    yp = paddle.to_tensor(y)
    yt = torch.from_numpy(y)

    i_yp = istft(yp, n_fft=n_fft, hop_length=hop_length, window=win_p)
    i_yt = torch.istft(yt, n_fft=n_fft, hop_length=hop_length, window=win_t)
    i_yl = librosa.istft(y.squeeze(0), hop_length=hop_length, window=win_l)

    print('*'*10, 'in dtype', '*'*10)
    print(f'paddle: {yp.dtype}, torch: {yt.dtype}, librosa: {y.dtype}')
    print('*'*10, 'out dtype', '*'*10)
    print(f'paddle: {i_yp.dtype}, torch: {i_yt.dtype}, librosa: {i_yl.dtype}')
    print()

    print('*'*10, 'in shape', '*'*10)
    print(f'paddle: {yp.shape}, torch: {yt.shape}, librosa: {y.shape}')
    print('*'*10, 'out shape', '*'*10)
    print(f'paddle: {i_yp.shape}, torch: {i_yt.shape}, librosa: {i_yl.shape}')
    print()

    print('*'*10, 'torch', '*'*10)
    for rtol in [1.e-7, 1.e-6, 1.e-5, 1.e-4, 1.e-3]:
        print(f'[paddle vs torch  ][rtol={rtol:.1E}]: {np.allclose(i_yp.numpy(), i_yt.numpy(), rtol=rtol)}')

    print('*'*10, 'librosa', '*'*10)
    for rtol in [1.e-7, 1.e-6, 1.e-5, 1.e-4, 1.e-3]:
        print(f'[paddle vs librosa][rtoj={rtol:.1E}]: {np.allclose(i_yp.numpy(), np.expand_dims(i_yl, 0), rtol=rtol)}')

    print('*'*10, 'torch vs librosa', '*'*10)
    for rtol in [1.e-7, 1.e-6, 1.e-5, 1.e-4, 1.e-3]:
        print(f'[torch  vs librosa][rtoj={rtol:.1E}]: {np.allclose(i_yt.numpy(), np.expand_dims(i_yl, 0), rtol=rtol)}')


if __name__ == '__main__':
    # ==== seed ====
    np.random.seed(0)

    # ==== data type ====
    switch_float_type(64)

    # ==== stft args ====
    attrs = {
        'shape': (1, 16000*30),
        'n_fft': 512,
        'hop_length': 256,
        'window': get_window('hann', Nx=512),
        # 'window': None,
        'device': 'cpu',
        # 'device': 'gpu:0',
    }

    # ==== test case ====
    yp, yt, yl = test_stft_api(**attrs)
    test_istft_api(yp.numpy(), **attrs)
```

**Comment by KPatr1ck** [comment]
# Test Summary

## STFT

- CPU, MKL, float64
  :white_check_mark: paddle vs torch
  :white_check_mark: paddle vs librosa

- CPU, PocketFFT, float64
  :white_check_mark: paddle vs torch
  :white_check_mark: paddle vs librosa

- GPU, CUDA, float64
  :white_check_mark: paddle vs torch
  :white_check_mark: paddle vs librosa

## ISTFT

- CPU, MKL, float64
  :white_check_mark: paddle vs torch
  :white_check_mark: paddle vs librosa

- CPU, PocketFFT, float64
  :white_check_mark: paddle vs torch
  :white_check_mark: paddle vs librosa
  
- GPU, CUDA, float64
  :white_check_mark: paddle vs torch
  :white_check_mark: paddle vs librosa

File: python/paddle/fluid/tests/unittests/test_frame_op.py [hunk 1]
```diff
@@ -21,23 +21,21 @@
 
 
 def frame_from_librosa(x, frame_length, hop_length, axis=-1):
-    if axis == -1 and not x.flags["F_CONTIGUOUS"]:
-        x = np.asfortranarray(x)
-    elif axis == 0 and not x.flags["C_CONTIGUOUS"]:
+    if axis == -1 and not x.flags["C_CONTIGUOUS"]:
```
Comment (iclementine) [inline python/paddle/fluid/tests/unittests/test_frame_op.py:24]: 这个地方 librosa 的细节确实有点奇怪

File: python/paddle/tensor/signal.py [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+         pad_mode: str='reflect',
+         normalized: bool=False,
+         onesided: bool=True,
+         name: Optional[str]=None):
+    '''
+        TODO(chenxiaojie06): Doc of stft.
+    '''
+    x_rank = len(x.shape)
+    assert x_rank in [1, 2], \
+        f'x should be a 1D or 2D real tensor, but got rank of x is {x_rank}'
```
Comment (iclementine) [inline python/paddle/tensor/signal.py:135]: fstring 要求 Python 版本 >= 3.7. 所以建议仍然使用 str.format 的写法。
paddle 框架内仍然暂时可以不写 typehint。（主要是框架的其他部分都没有 typehint。)

**Review by iclementine — APPROVED** [review]
LGTM

有一些修改意见可以参考修改一下，然后补全文档，拉取最新的代码解决冲突即可。

另外 stft 和 istft 的 python API，以及 frame_op 和 overlap_add_op 也需要单元测试。

File: python/paddle/tensor/signal.py [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+
+    assert x.dtype in [paddle.float16, paddle.float32, paddle.float64, paddle.complex64, paddle.complex128], \
```
Comment (iclementine) [inline python/paddle/tensor/signal.py]: 这里可以用框架内提供的函数 check_variable_and_dtype 进行数据类型的检查。

File: python/paddle/tensor/signal.py [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+        left = (n_fft - win_length) // 2
+        window_ = paddle.zeros(shape=(n_fft, ), dtype=x.dtype)
+        window_[left:left + win_length] = window
+
+    if center:
+        assert pad_mode in ['constant', 'reflect'], \
+            f'pad_mode should be "reflect" or "constant", but got "{pad_mode}".'
+
+        pad_length = n_fft // 2
+        x = paddle.nn.functional.pad(x.unsqueeze(-1),
```
Comment (iclementine) [inline python/paddle/tensor/signal.py:173]: 这里推荐使用 paddle.nn.functional.pad 的一个比较通用的版本。亦即，对于 nd tensor, 需要传一个 2n 的 List[int] 作为 padding_width, 分别是每一个维度的 start 和 end 的 padding 量。这个版本不用指定 data_format, 而且也不需要 squeeze unsqueeze 之类的操作就能使用。

File: python/paddle/tensor/signal.py [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+          return_complex: bool=False,
+          name: Optional[str]=None):
+    '''
+        TODO(chenxiaojie06): Doc of istft.
+    '''
+    x_rank = len(x.shape)
+    assert x_rank in [2, 3], \
+        f'x should be a 2D or 3D complex tensor, but got rank of x is {x_rank}'
+
+    assert x.dtype in [paddle.complex64, paddle.complex128], \
```
Comment (iclementine) [inline python/paddle/tensor/signal.py]: 同样可以用 check_variable_and_dtype

File: python/paddle/tensor/signal.py [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+        assert len(window.shape) == 1 and len(window) == win_length, \
+            f'expected a 1D window tensor of size equal to win_length({win_length}), but got window with shape {window.shape}.'
+    else:
+        window = paddle.ones(shape=(win_length, ))
+
+    window_ = window
+    if win_length < n_fft:
+        left = (n_fft - win_length) // 2
+        window_ = paddle.zeros(shape=(n_fft, ), dtype=window.dtype)
+        window_[left:left + win_length] = window
```
Comment (iclementine) [inline python/paddle/tensor/signal.py]: 此处何不用 pad?

File: python/paddle/tensor/signal.py [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+
+    x = x.transpose(
+        perm=[0, 2,
+              1])  # switch n_fft to last dim, egs: (batch, num_frames, n_fft)
+    norm = 'ortho' if normalized else 'backward'
+
+    # FIXME: Use ifft when it supports to call fft_c2r.
```
Comment (iclementine) [inline python/paddle/tensor/signal.py]: 现在应该已经支持了

File: python/paddle/fluid/tests/unittests/test_frame_op.py (hunk 1 above)
Comment (KPatr1ck) [reply python/paddle/fluid/tests/unittests/test_frame_op.py:24]: 修改后：axis -1 配合行连续，axis 0配合列连续，解决了3D以上出现的bug。

File: python/paddle/tensor/signal.py (hunk 2 above)
Comment (KPatr1ck) [reply python/paddle/tensor/signal.py:135]: 已去掉typehing。
已将fstring改为str.format写法。

File: python/paddle/tensor/signal.py (hunk 3 above)
Comment (KPatr1ck) [reply python/paddle/tensor/signal.py]: 已修改为check_variable_and_dtype

File: python/paddle/tensor/signal.py (hunk 5 above)
Comment (KPatr1ck) [reply python/paddle/tensor/signal.py]: 已修改为check_variable_and_dtype

File: python/paddle/tensor/signal.py (hunk 4 above)
Comment (KPatr1ck) [reply python/paddle/tensor/signal.py:173]: 只支持pad mode为constant，stft中需要用到不同的pad mode，暂先不改。

File: python/paddle/tensor/signal.py (hunk 6 above)
Comment (KPatr1ck) [reply python/paddle/tensor/signal.py]: 已修改为pad

File: python/paddle/tensor/signal.py (hunk 7 above)
Comment (KPatr1ck) [reply python/paddle/tensor/signal.py]: 保证功能正确的前提下，调用fft_r2c, fft_c2r, fft_c2c这三个api
